homeassistant/components/template/template_entity.py

```diff
@@ -430,3 +445,220 @@ async def async_added_to_hass(self) -> None:
     async def async_update(self) -> None:
         """Call for forced update."""
         self._async_update()
+
+
+class TemplateRestoreEntity(RestoreEntity, TemplateEntity):
+    """Template Entity that restores data."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Template Restore Entity init."""
+        super().__init__(*args, **kwargs)
+        self._restore: bool = False
+        self._save_state: bool = True
+        self._additional_data: list[str] = []
+
+    def add_template_attribute(
+        self, attribute: str, template: Template, *args, **kwargs
+    ):
+        """Store attribute to allow restore and setup template."""
+
+        super().add_template_attribute(attribute, template, *args, **kwargs)
+        if not template.is_static:
+            self.add_additional_data(attribute)
+
+    @callback
+    def _handle_results(self, *args, **kwargs) -> None:
+        # Do not save state if handle result fails.
+        self._save_state = False
+        super()._handle_results(*args, **kwargs)
+        self._save_state = True
+
+    @property
+    def restore(self) -> bool:
+        """Retrieve restore."""
+        return self._restore or False
+
+    @restore.setter
+    def restore(self, restore: bool) -> None:
+        """Set restore."""
+        self._restore = restore
+
+    @property
+    def save_state(self) -> bool:
+        """Retrieve save state."""
+        return self._save_state
+
+    @property
+    def additional_data(self) -> list[str]:
+        """Return additional data list."""
+        return self._additional_data or []
+
+    def add_additional_data(self, attribute: str) -> None:
+        """Add attribute to additional data list."""
+        if not hasattr(self, "_additional_data"):
+            self._additional_data = []
+
+        self._additional_data.append(attribute)
+
+    async def restore_entity(
+        self,
+    ) -> tuple[State | None, dict[str, Any] | None]:
+        """Restore the entity."""
+
+        if not self.restore:
+            return None, None
+
+        if (last_sensor_state := await self.async_get_last_state()) is None:
+            _LOGGER.debug("No state found to restore for entity %s", self.entity_id)
+            return None, None
+
+        # Restore all attributes.
+        _LOGGER.debug("Restoring entity %s", self.entity_id)
+
+        # Restore any attributes.
+        if self._attribute_templates is not None:
```

Comment:
#69350 takes a simpler approach; we know which attributes can be set by templates and which can't

Reply:
This was developed as a common solution for all the different template platforms without each having to make too many adjustments to support restore or having to identify exactly all values to be saved/restored.

Reply:
Using restore_attributes for attributes now as per #69350. Additional attributes that a entity requires are done seperately.